errors/error.go

```diff
@@ -4,78 +4,124 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
+
+	"github.com/gojek/fiber/protocol"
+	"google.golang.org/grpc/codes"
 )
 
-// HTTPError is used to capture the error resulting from a HTTP request
-type HTTPError struct {
+// FiberError is used to capture the error resulting from a HTTP request
+type FiberError struct {
 	Code    int    `json:"code"`
 	Message string `json:"error"`
 }
 
-// Error is a getter for the error message in a HTTPError object
-func (err HTTPError) Error() string {
+// Error is a getter for the error message in a FiberError object
+func (err FiberError) Error() string {
 	return err.Message
 }
 
-// ToJSON returns the HTTPError object as a Json encoded byte array
-func (err *HTTPError) ToJSON() ([]byte, error) {
+// ToJSON returns the FiberError object as a Json encoded byte array
+func (err *FiberError) ToJSON() ([]byte, error) {
 	return json.MarshalIndent(err, "", "  ")
 }
 
-// NewHTTPError returns an error of type HTTPError from the input error object.
+// NewFiberError returns an error of type FiberError from the input error object.
 // If the input error is already of the required type, it is returned as is.
 // If not, a generic request failed error is created from the given error.
-func NewHTTPError(err error) *HTTPError {
-	if httpErr, ok := err.(HTTPError); ok {
-		return &httpErr
+func NewFiberError(protocol protocol.Protocol, err error) *FiberError {
+	if fiberError, ok := err.(FiberError); ok {
+		return &fiberError
 	}
-	return ErrRequestFailed(err)
+	return ErrRequestFailed(protocol, err)
 }
 
 var (
-	// ErrRouterStrategyTimeoutExceeded is a HTTPError that's returned when
+	// ErrRouterStrategyTimeoutExceeded is a FiberError that's returned when
 	// the routing strategy fails to respond within given timeout
-	ErrRouterStrategyTimeoutExceeded = &HTTPError{
-		Code:    http.StatusInternalServerError,
-		Message: "fiber: routing strategy failed to respond within given timeout",
+	ErrRouterStrategyTimeoutExceeded = func(protocol protocol.Protocol) *FiberError {
+		statusCode := http.StatusInternalServerError
+		if protocol == "GRPC" {
+			statusCode = int(codes.Internal)
```

Comment:
I believe we should use DEADLINE_EXCEEDED for timeout? https://grpc.github.io/grpc/core/md_doc_statuscodes.html

Reply:
similar to the below, this is where `getRoutesOrder()` of the strategy timeout. `DEADLINE_EXCEEDED` (under ErrRequestTimeout) is returned when the route itself timeout.

I believe it was chosen as `InternalServerError` because it is mostly due to a faulty implementation of the `strategy`, although it is indeed timeout too. I can change this to `DEADLINE_EXCEEDED`, no strong opinion

Reply:
yeah, I think `DEADLINE_EXCEEDED` make more sense here.